server/app/services/program/predicate/PredicateGenerator.java

```diff
@@ -0,0 +1,290 @@
+                          .map(AndNode::create)
+                          .map(PredicateExpressionNode::create)
+                          .collect(toImmutableList()))),
+              predicateAction,
+              PredicateDefinition.PredicateFormat.OR_OF_SINGLE_LAYER_ANDS);
+        }
+
+      case SINGLE_QUESTION:
+        {
+          LeafOperationExpressionNode singleQuestionNode =
+              leafNodes.entries().stream().map(Map.Entry::getValue).findFirst().get();
+
+          return PredicateDefinition.create(
+              PredicateExpressionNode.create(singleQuestionNode),
+              predicateAction,
+              PredicateDefinition.PredicateFormat.SINGLE_QUESTION);
+        }
+
+      default:
+        {
+          throw new BadRequestException(
+              String.format("Unrecognized predicate format: %s", detectFormat(leafNodes)));
+        }
+    }
+  }
+
+  /**
+   * Gets the groupId and questionId for the provided predicate value matcher.
+   *
+   * @throws QuestionNotFoundException if the resulting questionId is not in the {@link
+   *     ReadOnlyQuestionService}
+   */
+  private static Pair<Integer, Long> getGroupIdAndQuestionId(
+      ReadOnlyQuestionService roQuestionService, Matcher matcher) throws QuestionNotFoundException {
+    int groupId = Integer.parseInt(matcher.group(1));
+    long questionId = Long.parseLong(matcher.group(2));
+
+    roQuestionService.getQuestionDefinition(questionId);
+
+    return Pair.of(groupId, questionId);
+  }
+
+  private static Pair<Scalar, Operator> getScalarAndOperator(
+      DynamicForm predicateForm, Long questionId) {
+    String rawScalarValue = predicateForm.get(String.format("question-%d-scalar", questionId));
+    String rawOperatorValue = predicateForm.get(String.format("question-%d-operator", questionId));
+
+    if (rawOperatorValue == null || rawScalarValue == null) {
+      throw new BadRequestException(
+          String.format(
+              "Missing scalar or operator for predicate update form: %s", predicateForm.rawData()));
+    }
+
+    try {
+      return Pair.of(Scalar.valueOf(rawScalarValue), Operator.valueOf(rawOperatorValue));
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(
+          String.format(
+              "Bad scalar or operator for predicate update form: %s", predicateForm.rawData()));
+    }
+  }
+
+  private static PredicateDefinition.PredicateFormat detectFormat(
+      Multimap<Integer, LeafOperationExpressionNode> leafNodes) {
+    if (leafNodes.size() > 1
+        || leafNodes.keySet().stream()
+            .map(leafNodes::get)
+            .filter(
+                (Collection<LeafOperationExpressionNode> leafNodeGroup) -> leafNodeGroup.size() > 1)
+            .findAny()
+            .isPresent()) {
+      return PredicateDefinition.PredicateFormat.OR_OF_SINGLE_LAYER_ANDS;
+    }
+
+    return PredicateDefinition.PredicateFormat.SINGLE_QUESTION;
```

Comment:
Curious: could this method just be 

`return leafNodes.values() > 1 ? OR_OF_SINGLE_LAYER_ANDS : SINGLE_QUESTION;`

Reply:
Yes! I forgot how multipmaps work.